File: problems/34-interval-scheduling/003-meeting-rooms-ii/java/Solution.java
import java.util.*;
public class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> ends = new PriorityQueue<>();
        int answer=0;
        for(int[] meeting:intervals){
            while(!ends.isEmpty() && ends.peek()<=meeting[0]) ends.poll();
            ends.add(meeting[1]); answer=Math.max(answer,ends.size());
        }
        return answer;
    }
}

File: problems/34-interval-scheduling/003-meeting-rooms-ii/python/solution.py
import heapq
class Solution:
    def min_meeting_rooms(self, intervals):
        intervals.sort()
        ends=[]
        answer=0
        for start,end in intervals:
            while ends and ends[0] <= start: heapq.heappop(ends)
            heapq.heappush(ends,end)
            answer=max(answer,len(ends))
        return answer
